### ISSUE TITLE
[Feature Request] `stat` and `lstat` APIs

### ISSUE THREAD

**Issue by sschwarzer** [racket/racket#3937]
**Is your feature request related to a problem? Please describe.**

I want to scan a directory tree to detect where two or more paths refer to the same inode, i. e. building a mapping inode -> list-of-paths. The `stat` system call (which seemingly doesn't have an API in current Racket) returns a struct that contains not only the inode number of a file or directory, but also the number of paths for the inode. So the scan would only need to store paths where the number of paths for an inode is greater than 1.

Being able to use stat would also be faster if you want different directory/file properties at once, so you don't have to make multiple indirect stat calls by calling, say, `file-size` and `file-or-directory-modify-seconds` after another.

**Describe the solution you'd like**

I'd like to see support for the [stat and lstat](https://man7.org/linux/man-pages/man2/stat.2.html) calls as a Racket API. The procedure, maybe called `stat`, would take a path and an optional keyword argument `#:follow-links?` and return a Racket `struct` similar to that of a the (l)stat(2) calls. I'm not sure about the name of the returned `struct`; `stat-result` may be too generic. Maybe `directory/file-info`?

Cross-platform considerations:

- Different platforms may define different fields for the stat (or equivalent call) result. I'm not sure if the result `struct` should have a union of all the fields that are possible on different platforms or if there should be different `stat` result `struct`s for different platforms (with the reasoning that a Racket process should only need one of the `struct` types).

- I'm not sure I understand the implications for distributed places though where you might want to transfer stat information between different machines on different platforms.

- On platforms where no `stat` or `stat`-alike call is supported, the `stat` call could raise an exception or return a partly populated `stat` result `struct` with the information that can be gathered by other API calls.

**Describe alternatives you've considered**

- There's a procedure [file-or-directory-identity](https://docs.racket-lang.org/reference/Filesystem.html#%28def._%28%28quote._~23~25kernel%29._file-or-directory-identity%29%29) which seems to contain inode information. But that still doesn't contain the number of file system entries for this directory/file, so while scanning I'd need to collect path information for _all_ directories/files, not only those where the number of paths for an inode is greater than 1. Besides that, the documentation for `file-or-directory-identity` is so vague that I'm not completely sure if I could rely on it for my purpose.

  Compared to `file-or-directory-identity` I would also welcome the device information in the `stat` result to detect if the file system tree crosses file system boundaries, which I'd like to avoid in my tool.

- I looked for a Racket package that implements the `stat` calls but didn't find any.

- I considered using the Racket foreign function interface to wrap the stat calls and provide a Racket API myself. However, in my opinion, the `stat` and `lstat` calls should be available in the standard library. Actually, I was quite surprised when I didn't find them. (Additionally, I asked in the Racket Slack in case I missed the procedures.) That said, if it's decided that the *stat calls won't be implemented in the Racket standard library, I might try to implement an external package that uses the FFI.

- In theory, it would be possible to call `stat(1)` via a `system*` call, but since I'm scanning a potentially large directory tree, that's not feasible.

- I might end up implementing my tool in Python or some other language which has the `*stat` calls. Actually, I started with a Python implementation.

**Do you want to contribute to this feature**

I had looked at the FFI documentation and was encouraged, but after having a look at [racket/src/io/file/main.rkt](https://github.com/racket/racket/blob/master/racket/src/io/file/main.rkt) and [racket/src/rktio/rktio_fs.c](https://github.com/racket/racket/blob/master/racket/src/rktio/rktio_fs.c) the potential implementation looks much more daunting to get right.

That said, I could contribute documentation and tests, at least under Linux. Since this would be my first Racket code contribution, I'd certainly have questions here and there though.

**Additional context**

- The stat support could be extended to the `fstat` call (also described in the above manpage), which works on a file descriptor instead of a string path.

- For comparison, here's an API description for the calls in Python (which has had `stat` and `lstat` for many years):

  https://docs.python.org/3/library/os.html#os.stat

  https://docs.python.org/3/library/os.html#os.stat_result

  Regarding an actual implementation I only could find https://github.com/python/cpython/blob/3.9/Modules/_stat.c , which doesn't seem helpful. (I wanted to check the code to see if it uses any platform-specific code.)

- As another reference, the documentation for Nim is here

  https://nim-lang.org/docs/os.html#getFileInfo%2Cstring
  
  https://nim-lang.org/docs/os.html#FileInfo
  
  Implementation: https://github.com/nim-lang/Nim/blob/44e653a9314e1b8503f0fa4a8a34c3380b26fff3/lib/pure/os.nim#L3184

**Comment by sschwarzer** [comment]
> I had looked at the FFI documentation and was encouraged, but after having a look at racket/src/io/file/main.rkt and racket/src/rktio/rktio_fs.c the potential implementation looks much more daunting to get right.

The problem isn't so much writing C code (although my C is quite rusty), but rather that I'd probably miss some details the implementation needs.

**Comment by sschwarzer** [comment]
By the way, one thing isn't clear to me when I look at the code [racket/src/rktio/rktio_fs.c](https://github.com/racket/racket/blob/master/racket/src/rktio/rktio_fs.c). I see a lot of `malloc` calls. Of course, I don't expect to see `free` calls since the just allocated data is returned and used "on the Racket side". But how/when is this data freed? Or are the C/Racket functions that use`malloc` and return the allocated data all leaking memory (i. e. even after garbage collection)?

**Comment by mflatt** [comment]
A `stat`-like interface does seem like a job for rktio instead of using the FFI directly. Even for just Unix variants, I'd expect some variation in exactly the layout of `stat` structures, and it's nice to be able to rely on the OS's C headers and provide a simple and consistent shape from rktio.

The `free` calls for rktio-allocated functions are on the Racket side, as you say. Look for calls to `rktio_free` in "racket/src/io/file/main.rkt" for CS, for example, or calls to just `free` in "racket/src/bc/file.c" for BC. In the CS implementation, you'll notice that the calls to `rktio_free` are in an atomic region along with the call to the allocating rktio function.

The  biggest part of the work here may be in sorting out the Windows side and figuring out what can be supported well and consistently enough. Historically, using C librar ywrappers like `_wstat` has not been a great choice, because it didn't handle certain kinds of paths or things like soft links, but I don't know these days.

**Comment by sschwarzer** [comment]
Finally I found the concrete implementation of CPython's low-level stat API. It's not in `_stat.c`, but here: [fileutils.c](https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Python/fileutils.c#L1076) . On Windows, they use [GetFileInformationByHandle](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfileinformationbyhandle) with [this struct](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/ns-fileapi-by_handle_file_information). In the Python implementation, you can see how they map the Windows struct data to the Posix struct data.

**Comment by sschwarzer** [comment]
For comparison, the Unix-like wrappers are described [here](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/stat-functions?view=msvc-160). This also gives some notes on the interpretation of the fields of the "returned" struct. It's not clear to me which of the `*stat*` functions are supported on which Windows versions.

**Comment by sschwarzer** [comment]
Today I looked into the behavior of soft links, hard links and junctions on Windows 10. For comparison, I was interested in how Python handles junctions when it comes to stat calls. I'm attaching my findings to this comment. (In the directory listing in `python_stat.txt`, you see a file junction. You can ignore it; I only wanted to see if Windows's `mklink` tool would create such a junction and it does, but you can't do anything useful with it.)

The results in `python_stat.txt` for soft links and hard links are what I would expect for Posix, too. Regarding junctions, `os.path.islink` reports `False` for junctions, which is different from Racket's [(link-exists? path)](https://docs.racket-lang.org/reference/Filesystem.html#%28def._%28%28quote._~23~25kernel%29._link-exists~3f%29%29).

Now for an API. I'd go with a single procedure `stat` with a required positional argument for the path and an optional positional argument that says whether the procedure should follow symlinks/soft links (corresponding to the C `stat` call) or not (corresponding to the C `lstat` call). I _guess_ we should also follow junctions in additon to regular soft links. The optional argument would be called `follow-links?` in the documentation and would default to `#t` (corresponding to the C `stat` call). I _think_ that's what most users would expect as default.

For the return value, I'm currently thinking of a regular Racket struct with at least the fields that are listed for the C stat result. From `man 2 stat` on my system:
```
           struct stat {
               dev_t     st_dev;         /* ID of device containing file */
               ino_t     st_ino;         /* Inode number */
               mode_t    st_mode;        /* File type and mode */
               nlink_t   st_nlink;       /* Number of hard links */
               uid_t     st_uid;         /* User ID of owner */
               gid_t     st_gid;         /* Group ID of owner */
               dev_t     st_rdev;        /* Device ID (if special file) */
               off_t     st_size;        /* Total size, in bytes */
               blksize_t st_blksize;     /* Block size for filesystem I/O */
               blkcnt_t  st_blocks;      /* Number of 512B blocks allocated */

               /* Since Linux 2.6, the kernel supports nanosecond
                  precision for the following timestamp fields.
                  For the details before Linux 2.6, see NOTES. */

               struct timespec st_atim;  /* Time of last access */
               struct timespec st_mtim;  /* Time of last modification */
               struct timespec st_ctim;  /* Time of last status change */

           #define st_atime st_atim.tv_sec      /* Backward compatibility */
           #define st_mtime st_mtim.tv_sec
           #define st_ctime st_ctim.tv_sec
           };
```
Do you think we should include any additional fields for Windows?

Fields which don't make sense for an OS should be set to `#f`. I think that makes more sense than having different struct fields for different operating systems because even for the same operating system, the availability of a field value might also depend on the type of file system where the queried path resides and in that case a user would have _two_ criteria / "coordinates" to think of. With the "missing" field value of `#f`, we'd be in trouble if we have/want a field that might be true or false, but this doesn't seem to apply.

I was thinking whether we could come up with a higher-level name than `stat`, but I think someone who wants this API, will look for this name. Also, to me the very concept of the `stat` call and its result seem relatively low-level, so that would make the name ok, I guess.

Please tell me what you think. Does the API fit into "typical" Racket idioms, especially the [filesystem APIs](https://docs.racket-lang.org/reference/Filesystem.html)? Anything else I overlooked? I'm sure there is. :-)

[python_stat.zip](https://github.com/racket/racket/files/6951302/python_stat.zip)

**Comment by mflatt** [comment]
Since the goal is to provide a way of getting information that's available from one system call, one thing to consider is how many OS calls you have to make on Windows to get the same information. I think it's at least two, but I'm not sure. If that's right, maybe that's fine, or maybe it makes sense to provide less information on Windows by default. Maybe there should be a way to say which information you want, and as many calls as necessary are made to get that information.

Instead of struct fields that are sometimes `#f`, this might be a better job for an immutable hash table. A hash table adapts much more nicely to future extensions, and the hash table could have keys for just the available information. A set of keys could be an optional argument to the function to indicate which things the function should try to include (even if it needs more system calls to get that information).

I'd call the function `file-or-directory-stat`. Using a long name has drawbacks, but it fits the current convention and would avoid potential conflicts with existing definitions of `stat` (more likely than existing definitions of `file-or-directory-stat`).

**Comment by sschwarzer** [comment]
@mflatt 
- struct vs. hash: Normally I'd say using low-level data structures in most APIs is a code smell, but in this case it would make sense, I guess. Although a struct would be extensible, too, a hash better fits the "open-ended" information we might need/want for the "stat" API.
- function name: I'm fine with `file-or-directory-stat`. I also tend to use longer names and since Racket doesn't use module name spaces for imported modules, longer names are more likely to avoid ambiguities, which just `stat` would most likely cause.

For the way forward, I suggest that I - with feedback from you or others if needed - work on documentation, tests and implementation (at least initially only for CS on Linux) "in parallel." I'd like to start with basic functionality though, for example, a single `stat(2)` call and the corresponding Racket side, and then refine/extend the documentation, tests and code iteratively, until we're there. I think that's better than writing, say, all of the documentation and then tests before starting with the implementation. Also, to be honest, I'd like to have some working code rather sooner than later because this would be more likely to keep me motivated. :-)

Regarding the tests, I looked/grepped through some of `racket/racket` and am not so sure where I'd put the tests. Would https://github.com/racket/racket/blob/master/pkgs/racket-test-core/tests/racket/file.rktl be the right place? (I hope not, since the file is already quite large and not very structured. ;-) )

Regarding the workflow: Should I only create a branch in my Github fork or also create a "WIP PR" from the start? I guess the latter would make review feedback more straightforward. I guess I'll have many questions until this ticket leads to production quality code. ;-)

What do you think?

**Comment by mflatt** [comment]
Yes, "file.rktl" is the right place, even though it's not so nice. Run with `racket -f file.rktl`.

Yes, a branch in your fork is right. A WIP PR sounds good.

**Comment by sschwarzer** [comment]
I started work on a [PR](https://github.com/racket/racket/pull/3961).

**Comment by sschwarzer** [comment]
So I need to change (relative to the Git working directory root):

- pkgs/racket-doc/scribblings/reference/filesystem.scrbl
- racket/src/rktio/rktio_fs.c
- racket/src/io/file/main.rkt
- pkgs/racket-test-core/tests/racket/file.rktl

You already told me on Slack how to rebuild the documentation, with `raco setup -l scribblings/reference` from the working directory root.

What minimum (i. e. relatively fast) steps do I need to rebuild/apply `racket/src/rktio/rktio_fs.c` and `racket/src/io/file/main.rkt` so that I can test the new code with `racket -f pkgs/racket-test-core/tests/racket/file.rktl`?

**Comment by mflatt** [comment]
For `racket/src/io/file/main.rkt`, you'll probably want to use `make demo` there and add to "demo.rkt". That way, you can try basic things without rebuilding Racket.

To build Racket with new primitives, start by changing the version number in `racket/src/version/racket_version.h`, and also in `pkg/base/info.rkt`.

In `racket/src/cs`, run `make` to get the updated `racket/src/io/file/main.rkt` compiled into `racket/src/cs/schemified/io.scm`. You will also need to add the new function to `racket/src/cs/primitive/kernel.ss`; the numbers you see there are arity masks in the sense of `procedure-arity-mark`.

Then `make` at the repo top level will build. You can use `make PKGS=""` to avoid installing and compiling the main-distribution packages. Probably "file.rktl" runs without having to install the `racket-test-core` package.

**Comment by sschwarzer** [comment]
This issue has been implemented with PR #3961 and can be closed.
